src/tools/rbd_wnbd/rbd_wnbd.cc

```diff
@@ -0,0 +1,1282 @@
+  ceph_assert(soft_disconnect_timeout >= 0);
+  ceph_assert(worker_count > 0);
+  int64_t timeout_ms = soft_disconnect_timeout * 1000;
+
+  Config cfg;
+  WNBDActiveDiskIterator iterator;
+  int err = 0, r;
+
+  boost::asio::thread_pool pool(worker_count);
+  LARGE_INTEGER start_t, counter_freq;
+  QueryPerformanceFrequency(&counter_freq);
+  QueryPerformanceCounter(&start_t);
+  while (iterator.get(&cfg)) {
+    boost::asio::post(pool,
+      [&, cfg]() mutable
+      {
+        LARGE_INTEGER curr_t, elapsed_ms;
+        QueryPerformanceCounter(&curr_t);
+        elapsed_ms.QuadPart = curr_t.QuadPart - start_t.QuadPart;
+        elapsed_ms.QuadPart *= 1000;
+        elapsed_ms.QuadPart /= counter_freq.QuadPart;
+
+        int64_t time_left_ms = max((int64_t)0, timeout_ms - elapsed_ms.QuadPart);
+
+        cfg.hard_disconnect = hard_disconnect || !time_left_ms;
+        cfg.hard_disconnect_fallback = true;
+        cfg.soft_disconnect_timeout = time_left_ms / 1000;
+
+        dout(5) << "Removing mapping: " << cfg.devpath
+                << ". Timeout: " << cfg.soft_disconnect_timeout
+                << "ms. Hard disconnect: " << cfg.hard_disconnect
+                << dendl;
+
+        r = do_unmap(&cfg, unregister);
+        if (r) {
+          err = r;
+          derr << "Could not remove mapping: " << cfg.devpath
+               << ". Error: " << r << dendl;
+        }
+        else {
+          dout(5) << "Successfully removed mapping: " << cfg.devpath << dendl;
+        }
+      });
+  }
+  pool.join();
+
+  r = iterator.get_error();
+  if (r) {
+    derr << "Could not fetch all mappings. Error: " << r << dendl;
+    err = r;
+  }
+
+  return err;
+}
+
+class RBDService : public ServiceBase {
+  private:
+    bool hard_disconnect;
+    int soft_disconnect_timeout;
+    int thread_count;
+
+  public:
+    RBDService(bool _hard_disconnect,
+               int _soft_disconnect_timeout,
+               int _thread_count)
+      : ServiceBase(g_ceph_context)
+      , hard_disconnect(_hard_disconnect)
+      , soft_disconnect_timeout(_soft_disconnect_timeout)
+      , thread_count(_thread_count)
+    {
+    }
+
+    int run_hook() override {
+      return restart_registered_mappings(thread_count);
+    }
+    // Invoked when the service is requested to stop.
+    int stop_hook() override {
+      return disconnect_all_mappings(
+        false, hard_disconnect, soft_disconnect_timeout, thread_count);
+    }
+    // Invoked when the system is shutting down.
+    int shutdown_hook() override {
+      return stop_hook();
+    }
+};
+
+static void usage()
+{
+  const char* usage_str =R"(
+Usage: rbd-wnbd [options] map <image-or-snap-spec>           Map an image to wnbd device
+                [options] unmap <device|image-or-snap-spec>  Unmap wnbd device
+                [options] list                               List mapped wnbd devices
+                [options] show-mapped <image-or-snap-spec>   Show mapped wnbd device
```

Comment:
Nit: the `-mapped` suffix seems like it's superfluous.

Reply:
I've updated the command but forgot about the help message, thanks.